### PR TITLE
4G or 5G timeseries in Dymola

### LINKED ISSUE #340: Templatize 5G Model for GMT
Jiazhen has reviewed & given feedback

Priority Order:
CHP - done
5th Gen
1st

-Work on couplings ongoing

- use MBL 9 (master branch of mbl)
- get `BuildingTimeSeriesWithETS` to run
    - copy our parameterized stuff into this new file
    - Add port connections from our old couplings to this building/ets combo
- Later build building models for Spawn & Teaser
-Will need upgrade to MBL v9 in order to be fully implemented
-Targeting completion in Q3 of FY '22
-Some software-side support needed for Nate--perhaps from Nick? Kyle to ID someone. 
-Everything appears to be working with master branch--cant be tested with GMT yet due to MBL version (blocked by MBL version). 

-Schema updates will be needed.

### LINKED ISSUE #486: Convert 4G partial building model to new v9
Still in progress. 

Jiazhen offering feedback on how to connect the instances to the district file.  Working on how to ensure DES model is built properly in both 4G/5G cases--close. 

Will need separate building models for 4G and 5G. 

.Replace existing 4G partial buildings with the one in MBL v9

The new one has the ambient, hot, and cold connections, as well as the results output.

image

Jiazhen reviewing, and working with Nate. Problem may be id'd--possible hiccup in redeclaration, between different versions of the time series model. 

Inconsistencies in display between text editor & Dymola itself.

### LINKED ISSUE #500: Support 4G and 5G model in same version of GMT
4G works in current develop, 5G is on a branch. Need both to be available.

### PR DESCRIPTION
**Pull request by vtnate**
#### Any background context you want to provide?
The existing 4G models haven't been updated to use MBLv9.1. There are new partial buildings, partial ets', and other new features we can take advantage of.

#### What does this PR accomplish?
- Updates existing 4G models to take advantage of MBLv9 capabilities
- Updates code to allow user selection of district generation (fourth or fifth are supported currently)
- Creates 5G TimeSeries model
    - ETS is embedded in the building model
    - District model has fewer objects
    - New couplings to connect new ambient loop to new ports on the building
#### How should this be manually tested?
- Create a 5G model:
    - `poetry run pytest tests/model_connectors/test_time_series_5g.py -m 'not simulation and not compilation'`
- Open the model in Dymola (MSLv4, MBLv9)
    - Confirm no errors on compiling (no errors when loading the model in Dymola)
- Simulate the model
    - May desire to change simulation to 365 days for more comprehensive test
    - Confirm no errors when simulating
#### What are the relevant tickets?
Resolves #340 
Resolves #486 
Resolves #500

### CODE REVIEW

**Comment by vtnate** [comment]
@nllong finish line

File: geojson_modelica_translator/model_connectors/load_connectors/time_series.py [hunk 1]
```diff
@@ -147,13 +155,24 @@ def to_modelica(self, scaffold):
...
 
-        self.run_template(
-            template=time_series_building_template,
-            save_file_name=os.path.join(b_modelica_path.files_dir, "building.mo"),
-            project_name=scaffold.project_name,
-            model_name=self.building_name,
-            data=combined_template_data
-        )
+        # This if statement exists only because we can't use the 5G model to run a 4G building.
+        if 'fifth_generation' not in building_template_data['district_type']:
```
Comment (nllong) [inline geojson_modelica_translator/model_connectors/load_connectors/time_series.py:159]: 👍

**Review by nllong — APPROVED** [review]
This looks good. model loads and run with msl 4.0 and mbl 9.1!

image

File: geojson_modelica_translator/model_connectors/couplings/coupling.py [hunk 2]
```diff
@@ -62,13 +62,17 @@ class Coupling(object):
...
         self._model_a = model_a
         self._model_b = model_b
+        self.district_type = district_type
         self._template_base_name = f'{model_a.model_name}_{model_b.model_name}'
 
-        self._template_dir = Path(__file__).parent / "templates" / self._template_base_name
+        if self.district_type == '5G':
+            self._template_dir = Path(__file__).parent / "5G_templates" / self._template_base_name
+        else:
+            self._template_dir = Path(__file__).parent / "templates" / self._template_base_name
```
[the remark this answers is not in the document]
Comment (vtnate) [reply geojson_modelica_translator/model_connectors/couplings/coupling.py:75]: Yes we absolutely should. I'll do that as a patch later, so we can get this out the door now.

File: geojson_modelica_translator/model_connectors/load_connectors/templates/TimeSeriesBuilding.mot [hunk 3]
```diff
@@ -1,76 +1,92 @@
 within {{ project_name }}.Loads.{{ model_name }};
-model building
+model BuildingTimeSeries
```
[the remark this answers is not in the document]
Comment (vtnate) [reply geojson_modelica_translator/model_connectors/load_connectors/templates/TimeSeriesBuilding.mot:2]: Ugh. Yes, absolutely. Thanks for catching that.

File: geojson_modelica_translator/model_connectors/districts/district.py [hunk 4]
```diff
@@ -122,6 +122,15 @@ def to_modelica(self):
                 'district_system': self.system_parameters.get_param('$.district_system')
             }
         }
+
+        if 'fifth_generation' in common_template_params['sys_params']['district_system']:
+            # 'bui' is how a 5G model refers to the 4G model while the templates build the model.
+            # 4G model is already self-sufficient so it needs that string to not exist.
+            # This is templated in TimeSeries component definitions, in the coupling templates
+            # FIXME: need a better variable name than 'is_5g_district' to be clearer in the template
+            common_template_params['globals']['is_5g_district'] = 'bui.'
```
[the remark this answers is not in the document]
Comment (vtnate) [reply geojson_modelica_translator/model_connectors/districts/district.py:131]: Turns out this is abandoned code from an earlier way of choosing 5G models. I'm removing it.

**Comment by vtnate** [comment]
@nllong 2 commits to address your comments. They are code cleanup, not substantive changes.

**Pull request merged**